res/controllers/midi-components-0.0.js

```diff
@@ -311,45 +308,34 @@
         shift: function() {
             this.inKey = "hotcue_" + this.number + "_clear";
         },
-        getColor: function() {
-            if (this.colorIdKey !== undefined) {
-                return color.predefinedColorFromId(engine.getValue(this.group, this.colorIdKey));
-            } else {
-                return null;
-            }
-        },
         output: function(value) {
             var outval = this.outValueScale(value);
             // WARNING: outputColor only handles hotcueColors
             // and there is no hotcueColor for turning the LED
             // off. So the `send()` function is responsible for turning the
             // actual LED off.
-            if (this.colorIdKey !== undefined && outval !== this.off) {
-                this.outputColor(engine.getValue(this.group, this.colorIdKey));
+            if (this.colorKey !== undefined && outval !== this.off) {
+                this.outputColor(engine.getValue(this.group, this.colorKey));
             } else {
                 this.send(outval);
             }
         },
-        outputColor: function(id) {
-            var color = this.colors[id];
-            if (color instanceof Array) {
-                if (color.length !== 3) {
-                    print("ERROR: invalid color array for id: " + id);
-                    return;
-                }
+        outputColor: function(colorCode) {
+            if (this.colorMapper !== undefined) {
+                var nearestColorValue = this.colorMapper.getNearestValue(colorCode);
+                this.send(nearestColorValue);
+            } else {
                 if (this.sendRGB === undefined) {
                     print("ERROR: no function defined for sending RGB colors");
                     return;
                 }
-                this.sendRGB(color);
-            } else if (typeof color === "number") {
-                this.send(color);
+                this.sendRGB(colorCodeToColor(colorCode));
```

Comment:
getNearestValue does not return an RGB color code. It returns the value (e. g. a midi byte) that is associated with a certain color.

Reply:
Ah, got it we have two send functions.
Can we rename the functions to match the values? something like sendColorValue(colorValue) and sendRGB(colorCodeToRGB())
What is colorCode in this game, btw.?

Reply:
> What is colorCode in this game, btw.?

e.g. 0xFF0000

Reply:
Ok, I renamed the functions and added some comments. I hope that clears it up.

Reply:
mm .. send() is the generic function uses for everything. 
Is nearestColorValue a midi command?

Reply:
You can map any `QScriptValue`/`QVariant` to a color. When you use our JS components, you're supposed to assign a Midi byte value to each color (so that `getValueForNearestColor` returns a midi byte). I did not name it MidiValue because HID controllers with a fixed color set might find the `ColorMapper` class useful, too.